### PR TITLE
Refactor InfoScanners

### PR DESCRIPTION
**Pull request by denizt**
We were not sharing any code between the scanners although they have some logic that can be easily shared. This patch is the initial refactoring to remedy the situation, however there is space left for more refactoring.

### CODE REVIEW

**Comment by Montellese** [comment]
Please take a look at #6406 which will remove the `CThread` part from `CVideoInfoScanner` as it will be executed as an asynchronous job.

File: xbmc/commons/InfoScanner.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ *  along with XBMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+#pragma once
+#include "utils/URIUtils.h"
+#include "filesystem/File.h"
+#include "threads/Thread.h"
+
+using namespace std;
```
Comment (Jalle19) [inline xbmc/commons/InfoScanner.h]: Don't put this in headers

**Comment by Memphiz** [comment]
thx @Jalle19 for seeing the evil before it happenend :)

**Comment by denizt** [comment]
@Montellese that is a great step forward! :) Any reason not to make MusicInfoScanner, the code is similar between those too and yet different and this could be a good excuse to clean this up.

I will push a follow up change but should I rebase on that patch?

File: xbmc/commons/InfoScanner.cpp [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ */
+
+#include "InfoScanner.h"
+#include "Util.h"
+#include "utils/log.h"
+#include "URL.h"
+
+using namespace std;
+
+namespace XbmcCommons
```
Comment (Jalle19) [inline xbmc/commons/InfoScanner.cpp]: I don't know what our conventions are, but if you add `using namespace XbmcCommons` you don't have to implement everything inside a namespace block, thus saving you one step of indentation.

File: xbmc/commons/InfoScanner.h [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with XBMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+#pragma once
```
Comment (Jalle19) [inline xbmc/commons/InfoScanner.h]: I am not absolutely sure but I think you need to have this directive at the first line in the file.

Comment (MartijnKaijser) [inline xbmc/commons/InfoScanner.h]: it doesn't really matter where it is as long as it's first before any actual code.

Comment (Paxxi) [inline xbmc/commons/InfoScanner.h]: Supposedly it's better for perf if it's first, some compilers will skip reading the file completely according to some talk from cppcon. I hardly think it matters though.

Comment (Montellese) [inline xbmc/commons/InfoScanner.h]: I think we have it at the very top in most header files (at least the ones I read frequently).

**Comment by Montellese** [comment]
@denizt: I just wanted to know if it actually works and I'm more comfortable with the video library. And there we have "Mark as watched/unwatched" which doesn't exist in the music library so it was easier to test. Ideally the same would be done for the music library but it can/should be done in a separate PR.

If you are going to work on this I would do it on top of the commits in my PR because then you get the changes I made to the job manager etc.

File: xbmc/commons/InfoScanner.h (hunk 3 above)
Comment (Paxxi) [inline xbmc/commons/InfoScanner.h]: Yup @Montellese we have it at the top in most but far from all and we even have it inside traditional ifdef guards :)

File: xbmc/commons/InfoScanner.cpp (hunk 2 above)
Comment (Paxxi) [inline xbmc/commons/InfoScanner.cpp]: don't encourage using namespace, that just drags everything back into the global namespace.

File: xbmc/music/infoscanner/MusicInfoScanner.cpp [hunk 4]
```diff
@@ -42,7 +42,6 @@
 #include "settings/Settings.h"
 #include "FileItem.h"
 #include "guilib/LocalizeStrings.h"
-#include "utils/StringUtils.h"
```
Comment (Paxxi) [inline xbmc/music/infoscanner/MusicInfoScanner.cpp]: why is this removed? StringUtils is still heavily used in the file.

**Comment by mkortstiege** [comment]
@denizt, just a quick fyi. https://github.com/xbmc/xbmc/pull/6406 has been merged and you can rebase on master now.

File: xbmc/commons/InfoScanner.h (hunk 3 above)
Comment (garbear) [inline xbmc/commons/InfoScanner.h]: putting #ifdef guards a the beginning of the header (theoretically) improves performance. `#pragma once` is processed differently and its location in the file won't affect performance

File: xbmc/music/infoscanner/MusicInfoScanner.cpp (hunk 4 above)
Comment (denizt) [inline xbmc/music/infoscanner/MusicInfoScanner.cpp]: Util.h includes StringUtils, thx for noticing, I removed it from VideoInfoScanner too.

Comment (Montellese) [inline xbmc/music/infoscanner/MusicInfoScanner.cpp]: IMO you shouldn't rely on headers being included by other headers. If you explicitly use `StringUtils` then also include `utils/StringUtils.h`. That way someone else can remove all usage of `CUtil` and also the `Util.h` header without any additional problems.

**Comment by denizt** [comment]
I just pushed an updated patch (hopefully) addressing all the above comments.

Re: pragma once Interesting discussion and I actually look the closest class (Exception) as a base to match the code style
Re: namespaces, same as above

@Montellese I am not very familiar with the codebase so doing things take some time at this stage. I will try to take a stab at it though :)

File: xbmc/music/infoscanner/MusicInfoScanner.cpp (hunk 4 above)
Comment (denizt) [inline xbmc/music/infoscanner/MusicInfoScanner.cpp]: Good point. I am all for removing/minimising Util classes, with your approach it would be easier.

Comment (denizt) [inline xbmc/music/infoscanner/MusicInfoScanner.cpp]: @Montellese please take a look at #6518 , it removes some occurrences where this was the case

**Comment by denizt** [comment]
Rebased on master and fixed explicit include of StringUtils.

Anything else in this patch that needs further work on?

File: xbmc/commons/InfoScanner.h [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+#include "utils/URIUtils.h"
+#include "filesystem/File.h"
+
+namespace XbmcCommons {
```
Comment (Montellese) [inline xbmc/commons/InfoScanner.h]: Please don't put this into `xbmc/commons` (or the `XbmcCommons` namespace) because that code is currently being built as a more or less separate library with very basic stuff in it (exceptions and logging) and the info scanner is nothing basic and very specific to Kodi.

File: xbmc/commons/InfoScanner.cpp (hunk 2 above)
Comment (Montellese) [inline xbmc/commons/InfoScanner.cpp]: @Paxxi: We are talking about the implementation of the class in that namespace here. IMO in that case a `using namespace` is OK. If you want to keep it this way at least change the indentation of the rest of the file to no base indentation.

File: xbmc/commons/InfoScanner.h [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  class InfoScanner {
+  public:
+    virtual bool DoScan(const std::string& strDirectory) = 0;
```
Comment (Montellese) [inline xbmc/commons/InfoScanner.h]: Please re-order the methods to
* constructor (if present)
* destructor
* virtual methods (if present)
* normal methods (if present)

and add some empty lines between the method definitions for better readability.

File: xbmc/commons/InfoScanner.h [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    virtual ~InfoScanner();
+    /*! \brief Check if the folder is excluded from scanning process
+     \param strDirectory Directory to scan
+     \param regexps Regular expression to exclude from the scan
+     \return true if there is a .nomedia file or one of the regexps is a match
+     */
+    bool IsExcluded(const std::string& strDirectory, const std::vector<std::string> &regexps);
+  private:
+    bool HasNoMedia(const std::string& strDirectory) const
```
Comment (Montellese) [inline xbmc/commons/InfoScanner.h]: Please move the implementation to the CPP file. Then you can also save the two includes (but you are missing includes for `<string>` and `<vector>`).

**Comment by Montellese** [comment]
I'll provide the VS sync one the issues I've brought up have been taken care of.

File: xbmc/commons/InfoScanner.h (hunk 5 above)
Comment (denizt) [inline xbmc/commons/InfoScanner.h]: Ok, any suggestions which namespace I should add this to?

File: xbmc/commons/InfoScanner.h (hunk 7 above)
Comment (denizt) [inline xbmc/commons/InfoScanner.h]: Will do, my intent here was to implicit-inline the method but that actually brings more complexity than its worth.

File: xbmc/commons/InfoScanner.h (hunk 5 above)
Comment (Montellese) [inline xbmc/commons/InfoScanner.h]: We don't really use namespaces that much so it doesn't have to be in any namespace. I'd move the files either into the `xbmc/` or the `xbmc/utils` directory.

**Comment by denizt** [comment]
@Montellese I addressed all your comments and rebased on master.

File: xbmc/InfoScanner.cpp [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+InfoScanner::~InfoScanner() {}
+
+bool InfoScanner::HasNoMedia(const std::string &strDirectory) const
+{
+  std::string noMediaFile = URIUtils::AddFileToFolder(strDirectory, ".nomedia");
+  return XFILE::CFile::Exists(noMediaFile);
+}
+
+bool InfoScanner::IsExcluded(const std::string& strDirectory, const std::vector<std::string> &regexps) {
```
Comment (Montellese) [inline xbmc/InfoScanner.cpp]: The curly brace goes on a new line.

File: xbmc/InfoScanner.cpp [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with XBMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "InfoScanner.h"
+#include "Util.h"
+#include "utils/log.h"
```
Comment (Montellese) [inline xbmc/InfoScanner.cpp]: Please sort your includes in a hierarchically alphabetic order.

File: xbmc/commons/Makefile [hunk 10]
```diff
@@ -1,6 +1,7 @@
 SRCS= \
 	Exception.cpp \
-	ilog.cpp
+	ilog.cpp \
+	InfoScanner.cpp
```
Comment (Montellese) [inline xbmc/commons/Makefile]: Wrong `Makefile`.

**Comment by denizt** [comment]
@Montellese PTAL

File: xbmc/InfoScanner.h [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+ *  You should have received a copy of the GNU General Public License
+ *  along with KODI; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+#pragma once
+#include <string>
+#include <vector>
+
+class InfoScanner {
```
Comment (Montellese) [inline xbmc/InfoScanner.h]: Curly brace on a new line (best make sure you do that for all of them).

File: xbmc/InfoScanner.cpp [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with KODI; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "filesystem/File.h"
```
Comment (Montellese) [inline xbmc/InfoScanner.cpp]: Please order them with the lowest hierarchy first and then by the sub-directories so
```cpp
#include "InfoScanner.h"
#include "URL.h"
#include "Util.h"
#include "filesystem/File.h"
#include "utils/log.h"
#include "utils/URIUtils.h"
```
(same goes for the includes in `xbmc/video/VideoInfoScanner.h`).

File: xbmc/InfoScanner.h (hunk 11 above)
Comment (denizt) [inline xbmc/InfoScanner.h]: I missed this one:( To avoid more ping-pongs with the style nits I am trying to configure a formatter following http://kodi.wiki/view/Official:Code_guidelines_and_formatting_conventions

Any other conventions the wiki may be missing?

**Comment by MartijnKaijser** [comment]
@denizt care to update this PR?

**Comment by denizt** [comment]
@MartijnKaijser rebased and followed-up @Montellese 's comments

**Comment by hudokkow** [comment]
jenkins build this please

**Comment by denizt** [comment]
It fails to build on Windows and AppleTv, is there a quick way to add the new files to those platforms' project files?

**Comment by mkortstiege** [comment]
appletv should be handling from within xcode. @Paxxi could you update the win32 project files?

**Comment by Paxxi** [comment]
@mkortstiege I'll sort the VS projects, make sure to remove the hdhomerun stuff from xcode proj files though.

**Comment by Paxxi** [comment]
cherry-pick Paxxi@805e79874f1a6bae12816615d5f2702f8e1d65a1 for VS projects

**Comment by mkortstiege** [comment]
@Paxxi thanks :)

**Comment by Paxxi** [comment]
just thought I'd mention it, InfoScanner doesn't really follow our naming rules, should be CInfoScanner but I'll leave it to you @mkortstiege to decide if it should be changed :)

**Comment by mkortstiege** [comment]
Jep, `CInfoScanner` then for the sake of consistency.

**Comment by denizt** [comment]
@mkortstiege Renamed to CInfoScanner and added the file to ATV Xcode. I merged @Paxxi's patch so the commit compiles.

**Comment by mkortstiege** [comment]
jenkins build this please

**Comment by denizt** [comment]
The ATV build still fails although I added the files to the project, or I thought I did... I thought we had stopped shipping Apple TV? Maybe @Memphiz can have a look?

**Comment by MartijnKaijser** [comment]
Memphiz will be busy with RL for some time. @denizt can you start with a rebase to master? @mkortstiege guess we'll need your assistance ;) http://jenkins.kodi.tv/job/IOS-ATV2/4554/console

**Comment by denizt** [comment]
@MartijnKaijser I rebased to master, no changes to Apple TV conf. :( I guess it will be taken off the build system some time soon anyway according to [this blog post](http://kodi.tv/farewell-to-future-appletv2-development/)?

**Comment by MartijnKaijser** [comment]
jenkins build this please

**Comment by MartijnKaijser** [comment]
@denizt seems the renaming of the files is gone

**Comment by denizt** [comment]
@MartijnKaijser I didn't rename the files, I renamed the classes, see https://github.com/denizt/xbmc/commit/37b64861c59ce3a62f9322bc03999dfb4e843072#diff-961ad40f45701d05e6b1351abad54e90R28

**Comment by Razzeee** [comment]
I would really love seeing this for Jarvis, can we get it up to speed again?

**Comment by denizt** [comment]
If we are no longer building for ATV2, I can rebase and update this PR. The only problem left was building on that platform.

**Comment by mkortstiege** [comment]
> If we are no longer building for ATV2, I can rebase and update this PR. The only problem left was building on that platform.

Jep, ATV2 was removed.

**Comment by Razzeee** [comment]
@denizt can you rebase this?

**Comment by denizt** [comment]
I wanted to rebase this but when I try to build the latest deps it fails at libsdl

```bash
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make[3]: *** [build/SDLMain.lo] Error 1
make[2]: *** [macosx10.11_x86_64-target/build/.libs/libSDL.a] Error 2
make[1]: *** [libsdl] Error 2
```
Is this a known issue, any workarounds?

**Comment by mkortstiege** [comment]
@denizt, sounds like dirty/broken depends.

**Comment by denizt** [comment]
@mkortstiege I did a clean build and still no dice. Are the Jenkins instances doing a clean build every time?

**Comment by Memphiz** [comment]
Jenkins builds as clean as we can get. I have no idea what sort of compile error you really have - the posted snipped is only part of the truth for sure because it could be a former error leading to this. Could you move your problem to the development forum and i can try to help you track it down ...

**Comment by denizt** [comment]
So I had some time to look at this issue today, turns out it's 10.11 SDK, using 10.10 SDK I was able to compile deps without problems. It looks like its a common issue, as a workaround I used the 10.10 SDK from Xcode 6.4 as mentioned in https://github.com/js-platform/node-webrtc/issues/235#issuecomment-152447773

**Comment by Memphiz** [comment]
mhh thats strange ... just built current master on osx 10.10 using xcode 7.2 and sdk 10.11 and what should i say:

Dependencies built successfully.

No issues ... will try your branch now ...

**Comment by Memphiz** [comment]
scratch that - won't compile your branch as it doesn't even touch depends. All in all i see no issues with 10.11 sdk and keep watching the forum for other issues popping up eventually ...

**Comment by denizt** [comment]
@Memphiz I will see if I can build on another El Capitan machine without problems next week

@mkortstiege I rebased, if all builds, should be good to go I guess?

**Comment by MartijnKaijser** [comment]
Let's see. jenkins build this please

**Comment by denizt** [comment]
@MartijnKaijser iOS build was failing as I didn't add the files to that target... I updated the commit and fixed that last issue

**Comment by MartijnKaijser** [comment]
jenkins build this please

**Comment by denizt** [comment]
@MartijnKaijser only Windows add ons build fails but I haven't touched it and it says non-blocking. I believe there it nothing left to do and the commit is good as as is? :smile:

File: xbmc/music/infoscanner/MusicInfoScanner.cpp [hunk 13]
```diff
@@ -439,8 +439,10 @@ bool CMusicInfoScanner::DoScan(const std::string& strDirectory)
   m_seenPaths.insert(strDirectory);
 
   // Discard all excluded files defined by m_musicExcludeRegExps
-  std::vector<std::string> regexps = g_advancedSettings.m_audioExcludeFromScanRegExps;
-  if (CUtil::ExcludeFileOrFolder(strDirectory, regexps))
+
```
Comment (Razzeee) [inline xbmc/music/infoscanner/MusicInfoScanner.cpp]: Unneeded line break

File: xbmc/music/infoscanner/MusicInfoScanner.h [hunk 14]
```diff
@@ -188,7 +189,7 @@ class CMusicInfoScanner : CThread, public IRunnable
   int GetPathHash(const CFileItemList &items, std::string &hash);
   void GetAlbumArtwork(long id, const CAlbum &artist);
 
-  bool DoScan(const std::string& strDirectory);
+  virtual bool DoScan(const std::string& strDirectory);
```
Comment (Razzeee) [inline xbmc/music/infoscanner/MusicInfoScanner.h]: I'm no expert on this, but should this really be virtual here? I would have thought we should loose the virtual and add an ```override```? But I may be out of my waters here.

Comment (Paxxi) [inline xbmc/music/infoscanner/MusicInfoScanner.h]: yup preferably it should be ```bool DoScan(const std::string& strDirectory) override;```
but it's not a blocker really

Comment (denizt) [inline xbmc/music/infoscanner/MusicInfoScanner.h]: Correct, let's be override-correct, push a new version rebased on master.

**Comment by Razzeee** [comment]
jenkins build this please
